docs/015_How-to-use-Git-as-a-database.md

```diff
@@ -0,0 +1,423 @@
+
+Git does not allow you to update the object if someone else has already changed it. So basically if you use Git to store your objects the only mechanism to avoid race conditions is optimistic locking.
+
+This way of using Git is like having an SQL table where you have version numbers for recording updates by default. That means every time you read a record you get the version number. When you update the record you check that the version number is still the same. The only difference compared with the normal SQL optimistic lock approach is that you always store the new version of the object, but you do not update the reference (pointer) to it. So you will continue retrieving the previous version.
+
+But what would happens if you try to push a reference with `git push` outside of the branch folder? Meaning using a custom reference like the one we have used before.
+
+The good news is that Git:
+
+- REJECTS updating any remote reference that refers to an object not in the local git object store.
+- REJECTS updating any remote reference that refers to a non-commit-object.
+- REJECTS updating any remote reference that changes a commit-object into referring a non commit-object.
+- REJECTS updating any remote reference that isn't a direct parent of the new-commit-object. (ie. non-fast-forward push).
+
+It only accepts fast forward `commit` reference updates.
+
+### How to design your unit of work
+
+Since you only can control concurrency at the object level (references) you have to make sure that you [design your aggregates](https://www.dddcommunity.org/library/vernon_2011/) in a way that each aggregate is a different reference (or branch if you use the second model).
+
+**What does that mean?**
+
+In the same example, you can have strict domain rules between counters. For example:
+
+- You cannot have more than a certain number of counters.
+- All counters cannot count more than 100.
+
+In those cases, you should have to define an object like a "pool of counters" and store all the values in the same object (branch).
+
+If we do not have any invariant between objects you can store each object in a different orphan branch. That way you can reduce conflicts and have a better performance. On the contrary, if you have some invariants with more than one object then you have to store the whole aggregate in the same branch, so you make sure we do not have any transactional inconsistencies. You could decide the storage method depending on the invariants declared in application.
+
+I recommend reading [these articles](https://www.dddcommunity.org/library/vernon_2011/) to know the trade-offs of aggregate design.
+
+### Atomic Git push
+
+Sometimes you will need to update the object but also to create other commits in others branches atomically. For example, if you want to update two object at the same time. You could want to update both of them or neither of them. That is to say, you could want to have transactions.
```

Comment:
ally. For example, if you want to update two object "s" at the same time. 

Missing and s

Reply:
Fixed